Handle invalid characters in an input file (issue #132)

rupertford opened the pull request.

rupertford commented:
I think this is ready for by by @arporter.

arporter commented on src/fparser/common/sourceinfo.py:
```diff
@@ -307,7 +308,11 @@ def get_source_info(file_candidate):
         #
-        with open(file_candidate, 'r') as file_object:
-            return get_source_info_str(file_object.read())
+        from fparser.common.utils import make_clean_tmpfile
```
Please update the docstring of the parent routine to follow rst markup.

rupertford replied:
OK (I think). File objects are different in py2 and py3.

arporter commented on src/fparser/common/readfortran.py:
```diff
@@ -1428,16 +1428,20 @@ def __init__(self, file_candidate, include_dirs=None, source_only=None,
         if isinstance(file_candidate, six.string_types):
             self.id = file_candidate
-            self.file = open(file_candidate, 'r')
+            from fparser.common.utils import make_clean_tmpfile
```
Please add a comment here (briefly) explaining what's going on.

rupertford replied:
OK, words added.

arporter commented on src/fparser/common/utils.py:
```diff
@@ -349,3 +349,60 @@ def __new__(metacls, name, bases, dict):
+    :returns: the name of the temporary file created by this function.
+    :rtype: str
+
+    :raises ParseError: if invalid input is found in the input file
```
It can raise InternalError too.

rupertford replied:
Well noticed. I forgot to update the documentation.

arporter commented on src/fparser/common/utils.py:
```diff
@@ -349,3 +349,60 @@ def __new__(metacls, name, bases, dict):
+                             "returned was {0}.".format(str(excinfo)))
+    orig_file.close()
+
+    # Tell codec to skip any errors
```
Nice. I thought we might actually have to do some work to throw-away things we didn't understand.

rupertford replied:
Yes, lucky!

arporter commented on src/fparser/common/utils.py:
```diff
@@ -349,3 +349,60 @@ def __new__(metacls, name, bases, dict):
+    except UnicodeDecodeError as excinfo:
+        if not skip_bad_input:
+            raise ParseError("Bad character found in input file. Error "
+                             "returned was {0}.".format(str(excinfo)))
```
I think it would be good to log any errors reading the file here. This is because we are otherwise silently modifying the contents of the file being parsed and that's not good - if we log it then the information will be reported should there be a subsequent problem.

arporter commented:
Having done the review I realised, on reflection, that we should probably log the fact we've had to throw away some characters. See in-line comment.

rupertford commented:
This is now waiting for pr #185 to go on master as the changes to the read.py and parse.py scripts in this PR relate the those in #185. Once merged we need to add a logging test for read.py and parse.py. We could also try to remove the replication of the log message (this is because the same file is read in two different places) but that might not make sense.

rupertford commented:
I think we're ready for the next review from @arporter. I've not added script tests for logging as the relevant section of code is already tested. I'm not 100% sure about this as I needed to add a logging setup call for the scripts to produce logging output, but I can't see how to test this anyway.

I've also left the two log warnings in, as both places call the same function that produces a log warning and it didn't seem to make sense to make it optional for one of them.

rupertford commented:
I'm not sure what's happened with codecov. It seems to be complaining about something that should be covered.

arporter commented on src/fparser/common/utils.py:
```diff
@@ -81,6 +82,7 @@
            'CHAR_BIT','str2stmt',
            'classes']
 
+import logging
```
Although not strictly part of this PR, pylint complains that the `import sys` below is not needed.

rupertford replied:
OK, I've fixed this (although there are many other pylint errors not related to this change in this file).

arporter commented on src/fparser/common/utils.py:
```diff
@@ -349,3 +351,65 @@ def __new__(metacls, name, bases, dict):
+        if not skip_bad_input:
+            raise ParseError("Bad "+message)
+        # Log the fact that this character will be removed from the input file
+        logging.getLogger(__name__).warning("Skipped bad "+message)
```
My pylint complains about this line:
    W1201: Specify string format arguments as logging function parameters (logging-not-lazy)

rupertford replied:
That's a weird one. You need to use %s format and pass the contents of %s as additional arguments in the logger. FIxed in any case.

arporter commented on src/fparser/scripts/fparser2.py:
```diff
@@ -68,6 +68,9 @@
 from __future__ import print_function
 from fparser.scripts.script_options import set_fparser_options
 
+import logging
```
pylint complains that this import is in the wrong place.

rupertford replied:
Quite right. Apologies.

arporter commented on src/fparser/scripts/parse.py:
```diff
@@ -65,6 +65,10 @@
 from fparser.scripts.script_options import set_parse_options
 import fparser.common.sourceinfo
+
+import logging
```
Again, pylint complains this import is in the wrong place.

rupertford replied:
Again, my mistake. Fixed.

arporter commented:
All looking good now apart from that missed line. AFAICS, we never test `scripts/parse.py` - we have tests for fparser2 scripts but not for fparser1. Hopefully you can just copy what we do for those...?
Apart from that, it's just a matter of fixing a few minor pylint errors/warnings.

rupertford commented:
> All looking good now apart from that missed line. AFAICS, we never test `scripts/parse.py` - we have tests for fparser2 scripts but not for fparser1. Hopefully you can just copy what we do for those...?
> Apart from that, it's just a matter of fixing a few minor pylint errors/warnings.

I've fixed the minor pylint errors/warnings/ Could you please explain which missing line you are referring to (sorry if it is obvious but I can't find it). I was also expecting to leave fparser1 because its classes have not been changed and this PR is not about fparser1 - I just added the logging configuration to the parse script so that it would output any logging warnings (like for the other scripts). If you think I should therefore test the whole of the parse.py script because of this change then I will do but it seems a little excessive.

rupertford commented:
Fixes and comments. Back to @arporter.

arporter commented:
All pylint issues have been addressed. However, in doing this you've broken a `print()` in the parse.py script. (You've removed the parentheses so it is now only valid in Python 2.) Please add a `from future import print_function` and put back the parentheses.
I'm not asking for full testing of the parse script since, as you say, it's an fparser1 thing. I just want a test to cover the addition of logging to the script (since it uses the now modified FileReader).

arporter commented:
I'm trying this branch of the parser with the AMM12 NEMO configuration and have hit a problem:

Traceback (most recent call last):
  File "/home/kbc59144/Projects/fparser/src/fparser/scripts/fparser2.py", line 125, in <module>
    main()  # pragma: no cover
  File "/home/kbc59144/Projects/fparser/src/fparser/scripts/fparser2.py", line 120, in main
    runner(parser, options, args)
  File "/home/kbc59144/Projects/fparser/src/fparser/scripts/fparser2.py", line 90, in runner
    reader = FortranFileReader(filename)
  File "/home/kbc59144/Projects/fparser/src/fparser/common/readfortran.py", line 1438, in __init__
    tmpfile = make_clean_tmpfile(file_candidate)
  File "/home/kbc59144/Projects/fparser/src/fparser/common/utils.py", line 410, in make_clean_tmpfile
    temp_file.write(orig_file.read())
UnicodeEncodeError: 'ascii' codec can't encode character u'\xb2' in position 25839: ordinal not in range(128)

arporter commented:
(This is with Python 3.) Have we read the file in in utf8 and are then trying to write it out as ASCII?

arporter commented:
The offending line is where someone has got all fancy with their comments:

               ! psi3+ : stable : B=-KhN²<0 => N²>0 if rn2>0 dir = 1 (stable) otherwise dir = 0 (unstable)

arporter commented:
With the exception of the file containing the line mentioned above, the fparser2 parsed and de-parsed NEMO source now builds and runs successfully (NEMO 3.6, AMM12, 4 MPI procs, gfortran and OpenMPI).

rupertford commented:
I'm seeing the error you report, but for Python2 not 3. I'm guessing you must have been using Python2 on Glados (assuming it was Glados you were building on)?

arporter commented:
Oh yes, you're right - it's Python 2 by default on Glados. For some reason I'd thought it was 3.

rupertford commented:
OK, ready for next review from @arporter.

arporter commented on src/fparser/one/tests/test_scripts.py:
```diff
@@ -0,0 +1,89 @@
+import pytest
+from fparser.scripts import parse
+
+# pytest.py script function runner()
```
What does this comment mean?

rupertford replied:
This is meant to indicate the type/group of tests that are below. I've tried to make it more clear. I'll understand if you think it should be removed (given that there are currently only two tests :-)).

arporter commented on src/fparser/one/tests/test_scripts.py:
```diff
@@ -0,0 +1,89 @@
+    parse.main()
+    # capture the output and check that the code has been output
+    stdout, _ = capsys.readouterr()
+    print( stdout)
```
pylint complains about the whitespace here.

rupertford replied:
Yes, sorry. I've fixed this - and other issues - to make it pylint compliant.

arporter commented on src/fparser/common/tests/utf.f90:
```diff
@@ -0,0 +1,4 @@
+program utf
```
Please add license/copyright.

rupertford replied:
OK, done. I've also cut down the example.

arporter commented on src/fparser/one/tests/bad_char.f90:
```diff
@@ -0,0 +1,55 @@
+MODULE dynadv_cen2
```
Please add license/copyright. Since this is straight out of NEMO we should be including their license text. We could also chop this example down.

rupertford replied:
I've cut the example down so it is no longer nemo code (apart from one character!) and added our licence/copyright info.

rupertford commented:
Ready for next review from @arporter

arporter commented:
All comments have been addressed. I've taken the liberty of fixing the copyright date in bad_char.f90.
Documentation builds OK and does not need updating since this is a bug fix.
